### ISSUE TITLE
Really Slow on super+tab alt+tab

### ISSUE THREAD

**Issue by Bicet** [rbreaves/kinto#48]
Edit:
I'm on a ubuntu 19.10 clean install.
Whenever I activate keyswap the computer become barely usable, at least from an UI point of view.
Closing a window takes a couple of seconds from the click.

Is there something I'm missing? 
As soon as I stop the service, the whole thing is back to normal. 

Original:
As soon as keyswap is active alt+tab and super+tab are less responsive, and at least for me unusable.
The problem is more evident when there're many windows/apps open.

**Comment by rbreaves** [comment]
Hmmmm, I am not sure what would be causing high cpu usage for you. Can you make sure you get back a numerical value on this command?

```
cat /tmp/kinto/caret
```

You and also display the entire log of your Kinto session via this.

```
journalctl --user-unit=keyswap.service -b
```

I just spun up my ubuntu 19.10 VM a moment ago and I am watching the cpu and ram usage and I am just not experiencing anything odd at the moment. If any app may cause a spike in cpu usage though I'd think it would be the browsers like Firefox or Chrome as I am polling the caret/cursor status pretty often, once every half second, but it really hasn't shown anything significant cpu wise to me. I can't tell the difference between Kinto running and it not.

**Comment by rbreaves** [comment]
Also are you on the master branch or dev branch? The master branch probably doesn't have all the code needed for app switching or in app tab switching.

You could try running the dev branch, even though it is in flux at times, as I am balancing what needs to be done in a way that works for multiple distros and DE's at the moment. The tab switching in it I believe works much more like macOS and tends to work pretty well under Gnome based distros. I also having work almost as well under xfce4, well enough to release it - even though cycling forwards or backwards is problematic on terminals, I can only manage one or the other.

KDE though is being far more problematic with Alt+Tab switching than the Gnome or xfce4 though. Even if I can trigger the app switcher to appear, the Tab key does not continue to be passed along.

**Comment by kyuumeitai** [comment]
Hi there.
Manjaro Gnome user here. Also noted the same. 
The result of 
`cat /tmp/kinto/caret
`is
`2`
The result of 
`journalctl --user-unit=keyswap.service -b
`is

```mar 10 06:46:20 misato bash[11073]: Running init command: gsettings set org.gnome.mutter overlay-key ''
mar 10 06:46:20 misato bash[11073]: Starting keyswap...
mar 10 06:46:20 misato bash[11073]: First window name: Google-chrome
mar 10 06:46:20 misato bash[11073]: chrome: Google-chrome
mar 10 06:46:55 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:47:01 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:48:17 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:48:17 misato bash[11073]: term: Gnome-terminal
mar 10 06:48:36 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:48:36 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:48:36 misato bash[11073]: chrome: Google-chrome
mar 10 06:49:00 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:01 misato bash[11073]: gui: Pamac-manager
mar 10 06:49:18 misato bash[11073]:      Gnome-tweaks
mar 10 06:49:28 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:28 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:28 misato bash[11073]: chrome: Google-chrome
mar 10 06:49:29 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:29 misato bash[11073]: gui: Pamac-manager
mar 10 06:49:36 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:36 misato bash[11073]: term: Gnome-terminal
mar 10 06:49:43 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:43 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:49:44 misato bash[11073]: chrome: Google-chrome
mar 10 06:49:57 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:50:00 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:50:31 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:50:31 misato bash[11073]: term: Gnome-terminal
mar 10 06:50:35 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:50:35 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:50:35 misato bash[11073]: chrome: Google-chrome
mar 10 06:51:10 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:51:10 misato bash[11073]: term: Gnome-terminal
mar 10 06:51:31 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:51:31 misato bash[11073]: gui: Gnome-tweaks
mar 10 06:51:32 misato bash[11073]: Internal error:   Could not resolve keysym XF86FullScreen
mar 10 06:51:32 misato bash[11073]: term: Gnome-terminal
```

**Comment by rbreaves** [comment]
Thanks @kyuumeitai, I’ll take another look at this.

**Comment by rbreaves** [comment]
The app_cycle works fine for me for some reason, but in the dev branch I am remapping the app_cycle to the F13 and F14 keys as I am making an assumption they are not typically used by most people, but it is also dev - so it is not definite that I will continue to map to those keys before the actual release and merge to master.

Those keys in dev are more of a placeholder until I reach parity with all the major distros and DE's I want and am currently supporting. So to sum it up, I can understand why there might be slowness under the dev branch for app switching, but there's been no real attempt to add app switching into the master branch just yet.

~/.xkb/symbols/mac_gui
```
...
partial xkb_symbols "mac_appcycle" {
    // Cycle App and In App Tab Switching
    key <FK13> { [ F13 ] };
    key <FK14> { [ F14 ] };
    replace key <TAB> {
        type[Group1]= "ONE_LEVEL_CMD",
        symbols[Group1]= [
            Tab,
            Tab,
            F14,
            F13,
            NoSymbol,
            NoSymbol
        ],
        actions[Group1]= [
            // Base
            NoAction(),
            // Shift
            NoAction(),
            // Control
            Redirect(key=<FK14>),
            // Control + Shift
            Redirect(key=<FK13>),
            // Super
            Redirect(key=<TAB>,mods=Control,clearmods=Mod4+Super),
            // Super + Shift
            Redirect(key=<TAB>,mods=Control,clearmods=Mod4+Super)
        ]
    };
};
...
```

**Comment by rbreaves** [comment]
@kyuumeitai I was able to duplicate the following error message and have suppressed it now in master as it has no impact at all.

```
Internal error:   Could not resolve keysym XF86FullScreen
```

I am not sure why the Alt+Tab switching is sometimes delayed while the service is running, it is very odd behavior that does not show up under most other distros I have been testing under.

I've also noticed the IBus input method is not actually running correctly either and have been trying to resolve that problem, but I don't have a solution for it yet.

**Comment by kyuumeitai** [comment]
Hi there. 
Running master and updated.

when I see the entire log `journalctl -b` I see the following:

```
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 31 with keysym 31 (keycode a).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 32 with keysym 32 (keycode b).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 33 with keysym 33 (keycode c).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 34 with keysym 34 (keycode d).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 35 with keysym 35 (keycode e).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 38 with keysym 38 (keycode 11).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 36 with keysym 36 (keycode f).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 39 with keysym 39 (keycode 12).
mar 13 21:37:34 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 37 with keysym 37 (keycode 10).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 31 with keysym 31 (keycode a).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 32 with keysym 32 (keycode b).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 33 with keysym 33 (keycode c).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 34 with keysym 34 (keycode d).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 35 with keysym 35 (keycode e).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 38 with keysym 38 (keycode 11).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 36 with keysym 36 (keycode f).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 39 with keysym 39 (keycode 12).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 37 with keysym 37 (keycode 10).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 31 with keysym 31 (keycode a).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 32 with keysym 32 (keycode b).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 33 with keysym 33 (keycode c).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 34 with keysym 34 (keycode d).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 35 with keysym 35 (keycode e).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 38 with keysym 38 (keycode 11).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 36 with keysym 36 (keycode f).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 39 with keysym 39 (keycode 12).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 37 with keysym 37 (keycode 10).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 31 with keysym 31 (keycode a).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 32 with keysym 32 (keycode b).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 33 with keysym 33 (keycode c).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 34 with keysym 34 (keycode d).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 35 with keysym 35 (keycode e).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 36 with keysym 36 (keycode f).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 38 with keysym 38 (keycode 11).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 39 with keysym 39 (keycode 12).
mar 13 21:37:35 misato gnome-shell[1134]: Window manager warning: Overwriting existing binding of keysym 37 with keysym 37 (keycode 10).

```

**Comment by rbreaves** [comment]
@kyuumeitai I don't think those warnings are really relevant, they're not errors - just indication that the xkbcomp and custom keymap had been applied and overwrote the existing bindings as it should, even though not all characters listed are being replaced/updated.

Right now the main thing I want to resolve in Manjaro Gnome is why IBus doesn't run fully. I don't think it should be the cause of slow Alt+Tab or Super+Tab, but it is a problem.

**Comment by mpereira** [comment]
I'm seeing the same on Ubuntu 18.04, kinto master branch. Alt-tab is very slow.

```
$ cat /tmp/kinto/caret
1
```

```
$ journalctl --user-unit=keyswap.service -b
Hint: You are currently not seeing messages from other users and the system.
      Users in groups 'adm', 'systemd-journal' can see all messages.
      Pass -q to turn off this notice.
-- Logs begin at Wed 2020-03-04 12:23:10 CET, end at Fri 2020-03-27 15:19:05 CET. --
Mar 27 15:16:12 AID-LL02075 systemd[3093]: Started keyswap for Terminal and GUI.
Mar 27 15:16:12 AID-LL02075 bash[28872]: Importing user_config.json...
Mar 27 15:16:12 AID-LL02075 bash[28872]: Data from user_config.json imported successfully.
Mar 27 15:16:12 AID-LL02075 bash[28872]: Running init command: gsettings set org.gnome.mutter overlay-key ''
Mar 27 15:16:12 AID-LL02075 bash[28872]: Starting keyswap...
Mar 27 15:16:12 AID-LL02075 bash[28872]: First window name: Gnome-terminal
Mar 27 15:16:12 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:16:27 AID-LL02075 bash[28872]: chrome: Google-chrome
Mar 27 15:16:29 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:16:30 AID-LL02075 bash[28872]: chrome: Google-chrome
Mar 27 15:16:36 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:16:38 AID-LL02075 bash[28872]: chrome: Google-chrome
Mar 27 15:17:45 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:17:47 AID-LL02075 bash[28872]: chrome: Google-chrome
Mar 27 15:17:48 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:17:51 AID-LL02075 bash[28872]: gui: Microsoft Teams - Preview
Mar 27 15:17:53 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:17:56 AID-LL02075 bash[28872]: chrome: Google-chrome
Mar 27 15:18:31 AID-LL02075 bash[28872]: term: Gnome-terminal
Mar 27 15:18:39 AID-LL02075 bash[28872]: chrome: Google-chrome
Mar 27 15:19:05 AID-LL02075 bash[28872]: term: Gnome-terminal
```

**Comment by rbreaves** [comment]
@mpereira Thanks for the feedback and mentioning which branch you on. I am looking into it now on Ubuntu 18.04 as well. I am not able to duplicate the issue yet, at least not a severe delay - maybe a half second or full second on either the app switcher appearing or disappearing. I noticed there's a seg fault/dump that sometimes occurs after you use chrome for some reason so going to look at that too.

The caret output appears to be good, I get concerned when people report 2 back to me and has made me think that IBus not being fully configured could be causing a hangup but perhaps not. There may be some other type of loop occurring I suppose that I do not understand yet.

**Comment by rbreaves** [comment]
@mpereira You could also try the dev branch, but I do not know that it would give you a different result. It would enable the Alt-Tab switching to occur on Cmd-Tab instead however. Sorta breaks tab switching under chrome too, but it does keep working Firefox.

**Comment by mpereira** [comment]
Hi @rbreaves,

Thanks for the suggestions. I just tried the latest dev branch at `c6605c68cfe63c32041b044cf5b10b7337ec6e8f` and the slowness is similar. The window switcher UI hangs for a second or so after releasing alt-tab.

Is there any other data that I can provide you to help understanding why this happens?

**Comment by rbreaves** [comment]
@mpereira not that I can think of at this time, however jonchun pointed out an error that I can reproduce under KDE that could make Kinto more stable once I’ve resolved it.

The logs provided though don’t indicate a BadWindow error though, so I’m not positive that it’ll fix this too, but we’ll see.

**Comment by rbreaves** [comment]
Some proper multithreading has been added to handle the caret status, so I'd be interested in knowing if this may resolve any slowness issues you two could have been having? Before I was simulating it with a while loop and half second timeouts but now it has its own thread. No idea if it could improve Cmd+Tab or not, but it certainly wouldn't hurt anything.

@mpereira @Bicet

**Comment by rbreaves** [comment]
An additional change was made to help the new shortcuts to become permanent under Gnome, I discovered they did not survive a reboot. @Bicet @kyuumeitai A bug in the current Ubuntu 19.10 version of Gnome I guess. Looks like they had the problem once before not too long ago.

Also I updated it some for KDE as well, KDE still has some difficult in releasing the Shift key after tabbing - tapping shift or enter does release it, but it is not an issue on any other DE I have come across.

I will be closing this ticket for now unless I hear back that this is still an issue. A lot of work has been done between the multi-threading and other improvements, so I hope this has been knocked out, but again just let me know.
